helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -697,10 +706,16 @@ void createClient() throws Exception {
     int retryCount = 0;
     while (retryCount < 3) {
       try {
-        // TODO: synchronize this block and wait for the new non-zero session ID updated.
-        _zkclient.waitUntilConnected(_connectionInitTimeout, TimeUnit.MILLISECONDS);
-        handleStateChanged(KeeperState.SyncConnected);
-        handleNewSession();
+        // TODO: might be better to create a new method to return session id in waitUntilConnected.
+        synchronized (this) {
+          if (!_zkclient.waitUntilConnected(_connectionInitTimeout, TimeUnit.MILLISECONDS)) {
```

Comment:
When waitUntilConnected return false, does it always mean it is timeout-ed (not other failures)?  Will this method throw out any other exceptions?  In general, we need to either catch all exceptions, or skip them to the upper-level calls, handle partial set of failure cases could be confusing.